DataScience/Analytics/prototype/landRegistry.py

```diff
@@ -0,0 +1,846 @@
+    :rtype: pandas.DataFrame
+    """
+    # merge to original information to the matched data
+    toMatch = toMatch.reset_index()
+    data = pd.merge(matches, toMatch, how='left', on='LR_Index')
+    data = pd.merge(data, AddressBase, how='left', on='AB_Index')
+
+    # drop unnecessary columns
+    if dropColumns:
+        data.drop(['LR_Index', 'AB_Index'], axis=1, inplace=True)
+
+    return data
+
+
+def checkPerformance(df, linkedData,
+                     prefix='LandRegistry', path='/Users/saminiemi/Projects/ONS/AddressIndex/data/'):
+    """
+    Check performance - calculate the match rate.
+
+    :param df: data frame with linked addresses and similarity metrics
+    :type df: pandas.DataFrame
+    :param linkedData: input edge case data, used to identify e.g. those addresses not linked
+    :type linkedData: pandas.DataFrame
+    :param prefix: prefix name for the output files
+    :type prefix: str
+    :param path: location where to store the output files
+    :type path: str
+
+    :return: None
+    """
+    # count the number of matches and number of edge cases
+    nmatched = len(df.index)
+    all = len(linkedData.index)
+
+    # how many were matched
+    print('\nMatched', nmatched, 'entries')
+    print('Total Match Fraction', round(nmatched / all * 100., 1), 'per cent')
+
+    # save matched
+    df.to_csv(path + prefix + '_matched.csv', index=False)
+
+    # find those without match
+    IDs = df['TransactionID'].values
+    missing_msk = ~linkedData['TransactionID'].isin(IDs)
+    missing = linkedData.loc[missing_msk]
+    missing.to_csv(path + prefix + '_matched_missing.csv', index=False)
+    print(len(missing.index), 'addresses were not linked...')
+
+
+def runAll():
+    """
+    Run all required steps.
+
+    :return: None
+    """
+    print('\nReading in Address Base Data...')
+    start = time.clock()
+    ab = loadAddressBaseData()
+    stop = time.clock()
+    print('finished in', round((stop - start), 1), 'seconds...')
+
+    print('\nReading in Land Registry data...')
+    start = time.clock()
+    testData = loadData()
+    stop = time.clock()
+    print('finished in', round((stop - start), 1), 'seconds...')
+
+    print('\nParsing address data...')
+    start = time.clock()
+    parsedAddresses = parseInputData(testData)
+    stop = time.clock()
+    print('finished in', round((stop - start), 1), 'seconds...')
+
+    # set index names - needed later for merging / duplicate removal
+    ab.index.name = 'AB_Index'
+    parsedAddresses.index.name = 'LR_Index'
+
+    # split to those with full postcode and no postcode - use different matching strategies
+    msk = parsedAddresses['Postcode'].isnull()
+    withPC = parsedAddresses.loc[~msk]
+    noPC = parsedAddresses.loc[msk]
+
+    print('\nMatching Land Registry to Address Base data...')
+    start = time.clock()
+    ms1 = ms2 = m1a = m1b = m2a = m2b = False
+    if len(withPC.index) > 0:
+        msk = withPC['BuildingNumber'].isnull()
```

Comment:
This is really horrible part of the code. One day I have enough time to make it nicer...